コメントをつけられるようにする　の提出

YuSa0-6 opened the pull request:
## 要件

- 本（books）のCRUDを作ったときの復習として日報（reports）が投稿できるようにする（reportsのCRUDを作る。入力項目はタイトルと本文だけで良い。`created_at`の日付を日報の投稿日とする）
- 日報を更新・削除できるのはその日報を投稿した本人のみ（本のCRUDは特に制限なし）
- 本と日報の各詳細画面からコメントを投稿できるようにする。
- コメントを削除できるのはそのコメントを投稿した本人のみ
- 投稿したコメントは本と日報の各詳細画面から確認できる。投稿した内容に加えて、投稿したユーザーの名前（未入力ならメアド）と投稿日時も表示する
- 本と日報の各詳細画面でコメントが投稿できている様子をスクショする
- rubocopとerblintをパスさせる（要スクリーンショット）

## 手順

- [x]  Railsアプリに reports（日報） の CRUD （一覧、個別、作成、編集、削除）の機能を追加する。
    - [x]  Reportモデルを作成
    - [x]  usersとreportsとの関係を設定する
- [x]  books と reports それぞれにコメントがつけられるようにする。
- [x]  作成したコメントを削除できるようにする（2023-02-12 追加）
- [x]  ポリモーフィック関連を使ってbookに付くコメントとreportに付くコメントを共通のコードで動くようにする。
- [x]  投稿者本人のみが削除・更新できるようにする

## 参考

[アクションディスパッチ[Routing](https://api.rubyonrails.org/classes/ActionDispatch/Routing.html)ポリモーフィックルート](https://api.rubyonrails.org/classes/ActionDispatch/Routing/PolymorphicRoutes.html#method-i-polymorphic_path)



[[【初心者向け】丁寧すぎるRails『アソシエーション』チュートリアル【幾ら何でも】【完璧にわかる】 - Qiita](https://qiita.com/kazukimatsumoto/items/14bdff681ec5ddac26d1#%E4%B8%AD%E9%96%93%E3%83%86%E3%83%BC%E3%83%96%E3%83%AB%E3%82%92%E4%BD%BF%E3%81%A3%E3%81%A6er%E5%9B%B3%E3%81%AE%E4%BD%9C%E6%88%90)](https://qiita.com/kazukimatsumoto/items/14bdff681ec5ddac26d1#中間テーブルを使ってer図の作成)

```mermaid
erDiagram
    
"Users" {
serial id PK
string name
string email
}
"Comments" {
serial id PK
integer user_id FK
integer commentable_id FK
string comment_type FK
string title
string body
}
"Reports" {
serial id PK
integer user_id FK
string title
string body
}
"Books"  {
serial id PK
string title
string memo
}
```

YuSa0-6 committed:
Reportモデルを作成

YuSa0-6 committed:
Revert "Reportモデルを作成"

This reverts commit b82c098ac8462e0f68a52eeabdf2b6ad2e1bb2d8.

YuSa0-6 committed:
migration file 追加

YuSa0-6 committed:
controllerとViewの雛形を作成

YuSa0-6 committed:
Reportのカラムを追加

YuSa0-6 committed:
ルーティング設定

YuSa0-6 committed:
ナビゲーションを追加とともに日本語に対応する

YuSa0-6 committed:
new page 作成

YuSa0-6 committed:
index page 作成

YuSa0-6 committed:
show page 作成

YuSa0-6 committed:
アソシエーションを設定するためにmdelに記述する

YuSa0-6 committed:
Revert "アソシエーションを設定するためにmdelに記述する"

This reverts commit 4801d96d8bb81b1ba3ee1702225df57cc06a62a6.

YuSa0-6 committed:
この日報を投稿したユーザーを取得するためにusers_controllerに追記

YuSa0-6 committed:
reports_controllerから日報を受け取るためにインスタンス変数に変更

YuSa0-6 committed:
reports_controllerを追加

YuSa0-6 committed:
comments tableを追加した。booksとreportsに同じ処理内容でコメントするためにポリモーフィックを設定した。

YuSa0-6 committed:
commentに対する関係を設定

YuSa0-6 committed:
commentsに対するルーティンを追加する。

YuSa0-6 committed:
comment form追加

YuSa0-6 committed:
books reports　各詳細画面にてformからコメントできるようにする

YuSa0-6 committed:
booksに投稿されたコメントを取得するため

YuSa0-6 committed:
並びが変わらないように`order(:id)`をつけて取得する

YuSa0-6 committed:
comments_controllerを追加

YuSa0-6 committed:
update destroy を追加する

YuSa0-6 committed:
books/ reports/にそれぞれコメントコントローラーを追記

YuSa0-6 committed:
削除時に関係のあるデータも削除するために追記

YuSa0-6 committed:
それぞれの詳細画面にコメントを表示する

YuSa0-6 committed:
コメント用のページを追記。コメントが自分の場合の時表示を変えられるようにするためにメゾットをす`user.rb`に追加

YuSa0-6 committed:
crud処理ができるようにルーティングを変更する

YuSa0-6 committed:
いらないものを削除

YuSa0-6 committed:
空白を消す

YuSa0-6 committed:
コメントの日本語を追加する

YuSa0-6 committed:
コメントなので削除と作成のみに絞る。

YuSa0-6 committed:
投稿者本人のみ　更新　削除ができるように、メゾットを定義する

YuSa0-6 committed:
コメントの削除ボタンを追加する

YuSa0-6 committed:
日報の作成、編集画面を追加。＿formにて`render`を使って共通にする

YuSa0-6 committed:
日報詳細画面にて投稿者本人にしか編集、削除ができないように表示を切り替える。

YuSa0-6 committed:
新規日報作成へのナビゲーションを追加。

YuSa0-6 committed:
使用しなかったので削除する

YuSa0-6 committed:
削除成功時の通知をi18nに対応する

YuSa0-6 committed:
Rubocopを通す

YuSa0-6 committed:
report indexページのタイピングミスを修正する。

YuSa0-6 commented on app/models/user.rb:
```diff
@@ -7,4 +7,8 @@ class User < ApplicationRecord
   has_one_attached :avatar do |attachable|
     attachable.variant :thumb, resize_to_limit: [150, 150]
   end
+
+  def name_or_email
+    name.empty? ? email : name
+  end
```
このメゾットは誰がコメントしたかなのでcomment.rbに記述する方が適しているかもしれない。

YuSa0-6 commented on app/views/comments/_comments.html.erb:
```diff
@@ -0,0 +1,22 @@
+  <strong><%= Comment.model_name.human %>:</strong>
+  <% if comments.any? %>
+    <ul>
+      <% comments.each do |comment| %>
```
N+1対策する必要あり。5/15

yoshitsugu replied:
ここも対応予定ですか？

YuSa0-6 commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,61 @@
+    return if @report.user_id == current_user.id
+
+    redirect_to report_path(@report.id)
+  end
```
ユーザー自身だった場合はなにもしない、そうでない場合は日報の詳細画面に飛ぶ。エラーメッセージ出す必要あり（？）

YuSa0-6 commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,61 @@
+  def create
+    @report = Report.new(report_params)
+    @report.user_id = current_user.id
+    @report.save
```
作成後にメッセージを出すか、作ったものに飛ぶようにする方が良い。

YuSa0-6 commented on app/views/comments/_comments.html.erb:
```diff
@@ -0,0 +1,22 @@
+      <% end %>
+    </ul>
+  <% else %>
+    (<%= 'no comments' %>)
```
i18nで対応する必要あり。

yoshitsugu replied:
ここはこれから対応予定ですか?

yoshitsugu commented on db/migrate/20240416004955_create_reports.rb:
```diff
@@ -0,0 +1,11 @@
+class CreateReports < ActiveRecord::Migration[7.0]
+  def change
+    create_table :reports do |t|
+      t.integer :user_id
```
これも `references` でよさそうな気がします。どうでしょうか。

YuSa0-6 replied:
そうですね。自動でIDを振ってくれるreferencesの方が良いと思います。
変更して、user_idで指定できるように変更しました。

yoshitsugu commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+class ReportsController < ApplicationController
+  before_action :set_report, only: %i[show edit update destroy]
```
**ここではこのままで大丈夫です**が、before_actionで値をセットするのはアプリケーションが複雑になってくるとつらくなってきます。
scaffoldで作られるものがこうなっていたと思いますし、チームの方針などによっては問題ないのですが、一応そういう問題があるということだけ認識おいてもらえればと思います！
参考: https://speakerdeck.com/shinkufencer/how-to-using-before-action-with-happy-in-rails

YuSa0-6 replied:
資料ありがとうございます。
扱いには気をつけたいですね。
ControllerかModelに処理を記述するように気をつけたいと思います。

yoshitsugu commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,61 @@
+    params.require(:report).permit(:title, :body)
+  end
+
+  def ensure_current_user
```
現在のユーザーであることをensureしてもしょうがない気がするので、 `ensure_owner` や `ensure_editor` とかにしたほうがいいかなと思いました。

yoshitsugu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,27 @@
+  end
+
+  def destroy
+    @comment = Comment.find(params[:id])
```
コメントも作成者しか消せないようにしましょう。

yoshitsugu commented on app/views/comments/_comments.html.erb:
```diff
@@ -0,0 +1,22 @@
+<div class="index-items">
+  <strong><%= Comment.model_name.human %>:</strong>
+  <% if comments.any? %>
+    <ul>
```
見た目の調整はこのプラクティスの範囲ではないのでできる範囲でいいですが、ちょっとコメントが右側に寄っているのが気になったので、横幅いっぱい使えるようにしたほうがいいかなと思いました。
![Screenshot from 2024-05-20 10-29-46](https://github.com/YuSa0-6/fjord-books_app-2023/assets/748445/bc34204c-8f41-4f45-adee-8a79c37c3baf)

YuSa0-6 replied:
![image](https://github.com/YuSa0-6/fjord-books_app-2023/assets/105704944/4461cf04-1c9e-47f6-b92e-72adbee3f4d9)
上記のように変更しました。
よろしくお願いします。

yoshitsugu reviewed (commented):
コメントしました！

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/5/files/f006ff13e17026dcc59b0a974964edec753a0d0f#r1606152118
t.integerからt.referencesに変更する
controllerにあった、user_idを指定する処理は不要なので削除する。

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/5#discussion_r1606152724
現在のユーザーではなくeditorやownerの方が適しているので、関数名を変更する。

YuSa0-6 committed:
reportsにuser_idを追加する。
migrationファイルを変更する。

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/5#discussion_r1606153765
ensure_editorメソッドのコードが読みづらかったので変更する。
コメントも日報と同じように投稿者じしんにしかレコードを操作できないようにするために, ensure_editorを追加する。

YuSa0-6 committed:
入力欄がcontentのところがtitleになっていたのを　contentに修正する。

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/5#discussion_r1606154392
コメントのスタイリングを変更する。

YuSa0-6 reviewed (commented):
コメントしました。

yoshitsugu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+class CommentsController < ApplicationController
+  before_action :ensure_editor, only: destroy
```
ここでエラーがでているようです。

<img width="768" alt="スクリーンショット 2024-05-22 202107" src="https://github.com/YuSa0-6/fjord-books_app-2023/assets/748445/19b92ee6-fc37-4c47-8a2c-9989b6ae4959">

yoshitsugu commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,59 @@
+    @report = Report.find(params[:id])
+    @report.destroy
+
+    respond_to do |format|
```
特にフォーマットによる表示分けなどをしない場合、respond_toは必要ありません。

yoshitsugu reviewed (commented):
コメントしました！

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/5#discussion_r1609776790
フォーマットなどで表示分けをしないのでrespond_toは必要ないため、削除する。

yoshitsugu commented on app/views/reports/_form.html.erb:
```diff
@@ -0,0 +1,15 @@
+  </div>
+
+  <div>
+    <%= form.label :content %>
```
気づくのが遅れましたが、ここが翻訳されていないのがちょっと気になりました。
![Screenshot from 2024-05-24 09-18-21](https://github.com/YuSa0-6/fjord-books_app-2023/assets/748445/1d23f7eb-10c1-4584-9123-f716704464e8)

yoshitsugu replied:
また、編集時に本文の内容が空になります。


https://github.com/YuSa0-6/fjord-books_app-2023/assets/748445/53548e92-f971-42f2-991e-7717a9bee2d6

yoshitsugu commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,57 @@
+
+  def update
+    if @report.update(report_params)
+      redirect_to @report, notice: t('controller.common.notice_update', name: Report.model_name.human)
```
ここの翻訳ができていなさそうです。
![Screenshot from 2024-05-24 09-19-35](https://github.com/YuSa0-6/fjord-books_app-2023/assets/748445/300722ef-3aea-4967-bdff-ecb3faaa870d)

yoshitsugu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,34 @@
+
+  def ensure_editor
+    @comment = Comment.find(params[:id])
+    redirect_to commentable_path(@comment.commentable_id) if @comment.user_id != current_user.id
```
作者じゃないコメントを消そうとするとここでエラーになりますね。
![Screenshot from 2024-05-24 09-23-53](https://github.com/YuSa0-6/fjord-books_app-2023/assets/748445/dc36f8d5-e8b0-472f-8055-30b9bc8558e8)

YuSa0-6 replied:
承知しました。
メッセージを出して対応しましたが、そもそも私が書いた`redirect_to`の処理自体はは必要でしたか？

https://github.com/YuSa0-6/fjord-books_app-2023/assets/105704944/25f9eac9-3495-4ed6-b23e-953f9c373fef

yoshitsugu replied:
> メッセージを出して対応しましたが、そもそも私が書いたredirect_toの処理自体はは必要でしたか？

どのような理由で「redirect_toの処理が不要かもしれない」と思いましたか？その判断理由を書いてもらわないと、こちらとしてもその考え方が妥当かどうかを判断できません。
こういった実装において、明示的に仕様として書いていない部分については基本的にあまりこれが正解、というものはありません。(これは実務でも同様です) 「今回はこういう判断でこういう実装にしました。」などを説明してもらわないと、判断の妥当性についてのアドバイスができないです。

YuSa0-6 replied:
私の考えとしては、削除できない->その日報ページにてその旨をメッセージに出すのが適していると思いこの実装にしました。
ですが、メッセージが出せるのならばredirect_toを使用しなくても他に良い方法があれば良いと思い上記の質問をしました。
説明不足でした。申し訳ありません。。。

yoshitsugu replied:
> ですが、メッセージが出せるのならばredirect_toを使用しなくても他に良い方法があれば良いと思い上記の質問をしました。

`redirect_to` 以外でメッセージを表示する方法はあるか？という質問であっていますか？
やるとしたらJSを使って非同期に確認する方法ですかね。
HTTPの基本を思いだしてもらうとよいのですが、同期でHTTPリクエストを送る場合、ステータスコード200などで返ってくるHTMLを表示する、もしくは302などでリダイレクトする、という動きになります。
このメッセージはHTML上でレンダリングされているメッセージなので、基本的にはredirect_toなどでリダイレクトした先でレンダリングされるか、リダイレクトなしでそのままレンダリングされるか、の2パターンくらいになります。
ですので、他の方法というとJSでの非同期通信くらいになるかな、という話です。回答になっているでしょうか。

YuSa0-6 replied:
ありがとうございます。参考になりました。疑問に思ったので質問させていただきました。
今回はこのままの実装で問題ないと考えています。

yoshitsugu reviewed (commented):
コメントしました。
再提出時の動作確認が甘いようです。再提出時は一通り動作確認するようにおねがいします。

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/5#discussion_r1612501673]
カラム名と＿Formの名前が違うことから、DBに保存されず、翻訳がされていなかったので、_form名前をcontent->bodyに変更する。

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/5#discussion_r1612503477
翻訳ができていなかったので、コントローラのTypoを修正する。

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/5#discussion_r1612505898
作者じゃないユーザーがコメントを消すとundefined methodが発生するため、redirect_toの記述を変更する。メッセージも書いておく。

YuSa0-6 committed:
関連付けを使用して、日報を作成するように修正する。

yoshitsugu reviewed (commented):
コメントしました。

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/5/files/f006ff13e17026dcc59b0a974964edec753a0d0f#r1601326357
i18n対応する。

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/5/files/f006ff13e17026dcc59b0a974964edec753a0d0f#diff-d1362e358cc8657c022cfc98851a57ee03e3790feb20ba00b8e3da772950286e
n+1対策のためpreloadを使用する。

yoshitsugu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,34 @@
+  end
+
+  def ensure_editor
+    @comment = Comment.find(params[:id])
```
一旦このままでOKとしますが、ここと18行目で2回コメントをロードしているのが無駄に思います。 
対した処理をやっていないですし、before_actionで分けずにそのままaction内に書いちゃったほうがよいかなと思いました。

yoshitsugu commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,56 @@
+  end
+
+  def ensure_editor
+    @report = Report.find(params[:id])
```
ここも2回ロードすることになるのが無駄に思いました。

yoshitsugu reviewed (approved):
OKです！
コメントしましたので今後の参考までに。
